test/unit/math/rev/mat/fun/adj_jac_apply_test.cpp

```diff
@@ -1,12 +1,1198 @@
+                                 const Eigen::VectorXd& y_adj) {
+    double d;
+    std::vector<double> vd(size_vd);
+    Eigen::Matrix<double, Eigen::Dynamic, 1> ed1(rows_ed1);
+    Eigen::Matrix<double, Eigen::Dynamic, Eigen::Dynamic> ed2(rows_ed2,
+                                                              cols_ed2);
+    Eigen::Matrix<double, 1, Eigen::Dynamic> ed3(cols_ed3);
+    Eigen::Matrix<double, 1, 1> ed4;
+    d = y_adj(0);
+    for (int i = 0; i < size_vd; i++)
+      vd[i] = y_adj(1 + i);
+    for (int i = 0; i < rows_ed1; i++)
+      ed1(i) = y_adj(1 + size_vd + i);
+    for (int i = 0; i < rows_ed2 * cols_ed2; i++)
+      ed2(i) = y_adj(1 + size_vd + rows_ed1 + i);
+    for (int i = 0; i < cols_ed3; i++)
+      ed3(i) = y_adj(1 + size_vd + rows_ed1 + rows_ed2 * cols_ed2 + i);
+    ed4(0) = y_adj(1 + size_vd + rows_ed1 + rows_ed2 * cols_ed2 + cols_ed3);
+    return std::make_tuple(d, vd, int(), ed1, ed2, std::vector<int>(), ed3,
+                           ed4);
+  }
+};
+
+TEST(AgradRev, test_pass_through_working_all_var_types) {
+  int size_vd = 5, rows_ed1 = 3, rows_ed2 = 2, cols_ed2 = 3, cols_ed3 = 4;
+  stan::math::var d = 1.0;
+  std::vector<stan::math::var> vd(size_vd, 1.0);
+  Eigen::Matrix<stan::math::var, Eigen::Dynamic, 1> ed1(rows_ed1);
+  Eigen::Matrix<stan::math::var, Eigen::Dynamic, Eigen::Dynamic> ed2(rows_ed2,
+                                                                     cols_ed2);
+  std::vector<int> vi(5, 5);
+  Eigen::Matrix<stan::math::var, 1, Eigen::Dynamic> ed3(cols_ed3);
+  Eigen::Matrix<stan::math::var, 1, 1> ed4;
+
+  for (int i = 0; i < size_vd; i++)
+    vd[i] = 1.0;
+
+  for (int i = 0; i < rows_ed1; i++)
+    ed1(i) = 1.0;
+
+  for (int i = 0; i < rows_ed2 * cols_ed2; i++)
+    ed2(i) = 1.0;
+
+  for (int i = 0; i < cols_ed3; i++)
+    ed3(i) = 1.0;
+
+  ed4(0) = 1.0;
+
+  Eigen::Matrix<stan::math::var, Eigen::Dynamic, 1> y
+      = stan::math::adj_jac_apply<CheckAdjointsPassingThrough>(
+          d, vd, 5, ed1, ed2, vi, ed3, ed4);
+
+  y(0).grad();
+  EXPECT_NEAR(y(0).val(), d.val(), 1e-10);
+  EXPECT_NEAR(d.adj(), 1.0, 1e-10);
+  for (int i = 0; i < size_vd; i++)
+    EXPECT_NEAR(vd[i].adj(), 0, 1e-10);
+  for (int i = 0; i < rows_ed1; i++)
+    EXPECT_NEAR(ed1(i).adj(), 0, 1e-10);
+  for (int i = 0; i < rows_ed2 * cols_ed2; i++)
+    EXPECT_NEAR(ed2(i).adj(), 0, 1e-10);
+  for (int i = 0; i < cols_ed3; i++)
+    EXPECT_NEAR(ed3(i).adj(), 0, 1e-10);
+  EXPECT_NEAR(ed4(0).adj(), 0, 1e-10);
+  stan::math::set_zero_all_adjoints();
+
+  for (int j = 0; j < size_vd; j++) {
+    y(1 + j).grad();
+    EXPECT_NEAR(y(1 + j).val(), vd[j].val(), 1e-10);
+    EXPECT_NEAR(vd[j].adj(), 1.0, 1e-10);
+    EXPECT_NEAR(d.adj(), 0.0, 1e-10);
+    for (int i = 0; i < size_vd; i++)
+      if (i != j)
+        EXPECT_NEAR(vd[i].adj(), 0, 1e-10);
+    for (int i = 0; i < rows_ed1; i++)
+      EXPECT_NEAR(ed1(i).adj(), 0, 1e-10);
+    for (int i = 0; i < rows_ed2 * cols_ed2; i++)
```

Comment:
Overall, I'd rather be testing with finite diffs than with hand-written derivatives like this.